go/libraries/doltcore/table/keyless_reader.go

```diff
@@ -17,25 +17,122 @@ package table
 import (
 	"context"
 	"fmt"
+	"io"
+
+	"github.com/dolthub/go-mysql-server/sql"
 
 	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/row"
 	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/sqlutil"
 	"github.com/dolthub/dolt/go/libraries/doltcore/table/typed/noms"
 	"github.com/dolthub/dolt/go/store/types"
 )
 
+type keylessTableReader struct {
+	iter types.MapIterator
+	sch  schema.Schema
+
+	// duplicates
+	row        row.Row
+	duplicates uint64
+}
+
+var _ SqlTableReader = &keylessTableReader{}
+
+// GetSchema implements the TableReader interface.
+func (rdr *keylessTableReader) GetSchema() schema.Schema {
+	return rdr.sch
+}
+
+// ReadSqlRow implements the SqlTableReader interface.
+func (rdr *keylessTableReader) ReadRow(ctx context.Context) (row.Row, error) {
+	if rdr.duplicates == 0 {
+		key, val, err := rdr.iter.Next(ctx)
+		if err != nil {
+			return nil, err
+		} else if key == nil {
+			return nil, io.EOF
+		}
+
+		rdr.row, rdr.duplicates, err = row.KeylessRowsFromTuples(key.(types.Tuple), val.(types.Tuple))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	rdr.duplicates -= 1
+
+	return rdr.row, nil
+}
+
+// ReadSqlRow implements the SqlTableReader interface.
+func (rdr *keylessTableReader) ReadSqlRow(ctx context.Context) (sql.Row, error) {
+	r, err := rdr.ReadRow(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return sqlutil.DoltRowToSqlRow(r, rdr.sch)
+}
+
 func newKeylessTableReader(ctx context.Context, tbl *doltdb.Table, sch schema.Schema, buffered bool) (SqlTableReader, error) {
-	return nil, fmt.Errorf("newKeylessTableReader is unimplemented")
+	rows, err := tbl.GetRowData(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var iter types.MapIterator
+	if buffered {
+		iter, err = rows.Iterator(ctx)
+	} else {
+		iter, err = rows.BufferedIterator(ctx)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &keylessTableReader{
+		iter: iter,
+		sch:  sch,
+	}, nil
 }
 
+// TODO: this is broken! (for partition boundaries that hit rows with cardinality > 1)
```

Comment:
The partition boundaries are in the noms layer, so I don't see how it's possible for a boundary to "hit" any row

Reply:
At the sql layer we partition rows by index. For keyless tables, map index != row index.
consider a table: 
```
row(1,1) x 1
row(2,2) x 1000
row(3,3) x 1
```
If we use map index to separate this table into three partitions, the middle partition will do significantly more work.

Reply:
It's ok for partitions to be uneven. That's not "broken", it's just suboptimal.